tests/src/Forms/ClientFormTest.php

```diff
@@ -4,16 +4,72 @@
 
 namespace SimpleSAML\Test\Module\oidc\Forms;
 
-use SimpleSAML\Module\oidc\Forms\ClientForm;
 use PHPUnit\Framework\TestCase;
+use SimpleSAML\Module\oidc\Forms\ClientForm;
+use SimpleSAML\Module\oidc\Forms\Controls\CsrfProtection;
+use SimpleSAML\Module\oidc\ModuleConfig;
 
 /**
  * @covers \SimpleSAML\Module\oidc\Forms\ClientForm
  */
 class ClientFormTest extends TestCase
 {
-    public function testIncomplete(): never
+    /** @var CsrfProtection */
+    protected CsrfProtection $csrfProtection;
+    /**
+     */
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $_SERVER['REQUEST_URI'] = '/';
+        $this->csrfProtection =  $this->getMockBuilder(CsrfProtection::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+    }
+
+    public static function validateOriginProvider(): array
     {
-        $this->markTestIncomplete();
+        return [
+            ['https://example.com', true],
+            ['https://example.com.', true],
+            ['http://foo.', true],
+            ['http://foo', true],
+            ['https://user:pass@example.com', false],
+            ['http://example.com', true],
+            ['https://example.com:2020', true],
+            ['https://localhost:2020', true],
+            ['http://localhost:2020', true],
+            ['http://localhost', true],
+            ['https://example.com/path', false],
+            ['https://example.com:8080/path', false],
+            ['http://*.example.com', true],
+            ['http://*.example.com.', true],
+            ['https://foo.example.com:80', true],
+            ['http://*.example', false],
+            ['http://foo.*.test.com', false],
+            ['http://*', false],
+            ['http://*.com', false],
+            ['https://test........', false],
+            ['https://developer.mozilla.org:80', true],
+            ['http://attacker.bar/test.php', false],
+            ['https://cors-test.codehappy.dev', true],
+            ['https://cors-test.codehappy.dev:8080', true],
+        ];
```

Comment:
Can you add a test that having just a plain domain name 'example.com' is not valid, so we confirm the protocol is always required.

Reply:
added